### PR TITLE
Offline/Slashing improvements

### PR DESCRIPTION
**Pull request by gavofyork**
- Avoid punishing the entire validator set - this doesn't make any sense and would only lead to catastrophic loss of funds in the case of a systematic flaw in the codebase.
- Maintain a history of offline validators.

### CODE REVIEW

File: srml/aura/src/lib.rs [hunk 1]
```diff
@@ -179,15 +179,19 @@ impl AuraReport {
...
 		let iter = (start_slot..validator_count).into_iter()
 			.chain(0..start_slot)
 			.enumerate();
 
+		// Don't include skipped_all in the bad block count, since that would imply the
+		// network as a whole was down, indicating a systematic issue rather than a
+		// minority of bad actors.
 		for (rel_index, actual_index) in iter {
-			let slash_count = skipped_all + if rel_index < skipped_after {
+			let slash_count = if rel_index < skipped_after {
```
Comment (rphmeier) [inline srml/aura/src/lib.rs]: applying the above logic, I don't think we should slash at all if everyone is going to be slashed. Clearly there has been some kind of network malfunction, so I don't think slashing those who were offline a little more is helpful either.

**Comment by gavofyork** [comment]
@rphmeier logic changed accordingly. Worth checking though.

File: srml/aura/src/lib.rs [hunk 2]
```diff
@@ -175,28 +175,29 @@ impl AuraReport {
...
+		// unfulfilling their specific duties. In this case, it doesn't make
+		// sense to punish anyone, so we guard against it.
+		if self.skipped < validator_count {
+
+			let start_slot = self.start_slot % validator_count;
+
+			// the number of validators whose turned were skipped.
+			let iter = (start_slot..validator_count).into_iter()
+				.chain(0..start_slot)
+				.enumerate();
```
Comment (andresilva) [inline srml/aura/src/lib.rs]:
```suggestion
				.take(self.skipped);
```

File: srml/aura/src/lib.rs [hunk 3]
```diff
@@ -175,28 +175,29 @@ impl AuraReport {
...
+
+			for (rel_index, actual_index) in iter {
```
Comment (andresilva) [inline srml/aura/src/lib.rs]:
```rust
for index in iter {
	punish_with(index, 1);
}
```

**Pull request merged**